articles/testing-documentation-snippets/article.md

```diff
@@ -0,0 +1,173 @@
+
+
+For writing these tests, we opted for a YAML file in front of the snippets files. It simple to write, simple to read and simple to parse. They have to describe the test, give an expected string and the template we want to use. We also imagined a system of hooks (before and after) like in every test framework to execute some script for cleaning or other operations.
+
+```yaml
+name: kuzzle-connect
+description: Connects the SDK to Kuzzle
+hooks:
+  before: 
+  after: 
+template: default
+expected: Successfully connected
+```
+
+I found a cool Node.js package on NPM named [Nexpect](https://github.com/nodejitsu/nexpect). Then, I wrote a Node.js script for running tests for each language. For an easy implementation of every languages I make an abstract Tester class and I implement a children class for each language with lint and run command. In this way if we want to add another language for our SDKs it's very easy to just create a class and extend it to the mother tester class and only override commands to be executed for this new language.
+
+Here, the abstract class
+
+```js
+class Tester {
+
+  constructor () {
+    if (new.target === Tester) {
+      throw new TypeError("Cannot construct Tester instances directly");
+    }
+  }
+  
+  execLintCommand () {
+    // return a promise will executing 
+    // this.lintCommand added in childrens class
+  }
+  
+  execRunCommand () {
+    // return a promise will executing 
+    // this.runCommand added in childrens class
+  }
+  
+  // [...]
+}
+```
+
+And Here, the extended class for testing snippets in Go
+
+```js
+class GoTester extends Tester {
+  
+  constructor() {
+    super();
+    this.language = 'go';
+    this.runCommand = 'go run';
+    this.lintCommand = 'golint';
+  }
+
+}
+```
+
+And finally the last schema with the full process
+
+![schema4](./images/schema4.png)
+
+## Continuous integration and validation
+
+The last step and not the least was to put all of this process in a CI/CD workflow. At Kuzzle we use a Git-flow process. Each change in our documentation leads to a pull-request. And for each pull-request, we use Travis CI to run all containers for each language. I also add a report system to easily see which test was failed.
+
+![reports](./images/reports.png)
+
+Additionally, Alex our team's devOps guy(aka "AWS lambda gardener") developed a tool in Go to automatically add a comment to github in each pull-request to add a summary of the tests reports, which makes it even easier to detect a problem in our documentation before send it to production
+
+![kuttlefish](./images/kuttlefish.png)
+
+[here](https://github.com/alexandrebouthinon/kuttlefish) a link to the repository of this tool (named "Kuttlefish").
+
+
+## Conclusion
+
+In conclusion, it was a great experience! And of course, not only mine. All the Kuzzle team was very implicated in the development of our new documentation system. We made many meetings to deliberate to the better way to implement all of this process.
```

Comment:
"All the Kuzzle team was very implicated" :point_right: "All the Kuzzle team got involved"

"We made many meetings to deliberate to the better way" :point_right: "We made a lot of meetings to decide what was the beat way"